src/template-expr.adoc

```diff
@@ -0,0 +1,255 @@
+
+
+==== Preventing Evaluation
+
+===== `*literal*`
+
+[{nrm}]
+----
+(*literal* _datum_)
+----
+
+Produces _datum_ as-is, preventing the operand from being evaluated as a template.
+
+For example, `(*literal* [1, (values 2 3), 4])` produces `[1, (values 2 3), 4]`; both the list
+and the S-expression are treated as literal, constant data, not as template expressions to be
+expanded.
+
+
+==== Conditionals
+
+These special forms allow output to vary based on whether a template produces zero, one, or more
+values.
+
+
+===== `*if_void*`
+
+[{nrm}]
+----
+(*if_void* _template1_ _template2_ _template3_)
+----
+
+Evaluates templates conditionally based on cardinality of a stream.
+
+The _template1_ is expanded to see if it produces any values.
+If and only if it produces no values, then _template2_ is expanded and its results returned.
+Otherwise, _template3_ is expanded and its results returned.
+
+
+===== `*if_single*`
+
+[{nrm}]
+----
+(*if_single* _template1_ _template2_ _template3_)
+----
+
+Like `*if_void*`, but expands _template2_ if and only if _template1_ produces exactly one value,
+otherwise expands _template3_.
+
+
+===== `*if_many*`
+
+[{nrm}]
+----
+(*if_many* _template1_ _template2_ _template3_)
+----
+
+Like `*if_void*`, but expands _template2_ if and only if _template1_ produces more than one
+value, otherwise expands _template3_.
+
+[{nrm}]
+----
+(*macro* decimal_constraint
+    [(precision *int{asterisk}*), (exponent *int{asterisk}*)]
+    {
+        precision: (*if_many* precision range::[precision] precision),
+        exponent:  (*if_many* exponent  range::[exponent]  exponent),
+    })
+----
+
+[{nrm}]
+----
+(:decimal_constraint (3) (-1))     ⇒ { precision: 3, exponent: -1 }
+(:decimal_constraint (1 5) (-5 0)) ⇒ { precision: range::[1, 5],
+                                       exponent: range::[-5, 0] }
+(:decimal_constraint (:) (3 max))  ⇒ { exponent: range::[3, max] }
+(:decimal_constraint (1) (:))      ⇒ { precision: 1 }
+----
+
+
+==== Mapping
+
+These special forms produce repeated output mapped across elements of a stream.
+
+===== `*for*`
+
+[{nrm}]
+----
+(*for* [(_id_ _in-template_), ...] _body-template_)
+----
+
+Iteratively expands the _body-template_ using individual values from the _in-templates_.
+
+Each iteration takes the next value from each _in-template_ stream; iteration stops when any
+stream ends.
+Local variables are created for each identifier _id_, bound to the current value from their stream.
+The _body-template_ is then expanded in that environment, and iteration proceeds.
+The result of the `*for*` expression is the concatenated results of the body expansions.
+
+NOTE: The termination rule is under discussion; see
+https://github.com/amazon-ion/ion-docs/issues/201
+
+
+=== Macro Invocation
+
+A macro definition can express its output in terms of other macros. Quite often, these will be
+macros provided by the Ion implementation,
+// —everything in System Macros 2023-05 is available by default—
```

Comment:
I think this is an internal doc leftover.

Reply:
It is, but I want to resurrect it with a proper link (the TODO on L205)